cli/lsp/diagnostics.rs

```diff
@@ -517,42 +517,79 @@ async fn generate_deps_diagnostics(
   .unwrap()
 }
 
+/// Converts relative paths of ignored files/folders to Url format.
+fn get_ignore_specifiers(
+  root: &ModuleSpecifier,
+  ignore: &[String],
+) -> Vec<ModuleSpecifier> {
+  // FIXME: `root_uri` from `ConfigSnapshot` are without a trailing slash,
+  // so `Url::join` treats the root as a file, not a directory, and erases the folder name.
+  // To fix that behaviour we just parsing `root_uri` again.
+  let root =
+    ModuleSpecifier::from_directory_path(root.to_file_path().unwrap()).unwrap();
+  ignore.iter().filter_map(|i| root.join(i).ok()).collect()
+}
+
 /// Publishes diagnostics to the client.
 async fn publish_diagnostics(
```

Comment:
I think maybe doing this check at `publish_diagnostics` is too late. Perhaps we should ignore these files in `generate_lint_diagnostics` so that we don't bother doing this work when not needed?

Reply:
Thanks! You're right, I thought it should ignore all the diagnostics.
So, if I get You right, I need to filter a result of `snapshot.documents.documents(true, true)` in `generate_lint_diagnostics` and leave the rest of the function as is.

Reply:
Yes, I believe that would be better. Also, it looks like we're not handling `files.include` in addition to `files.exclude`.

Reply:
> I think maybe doing this check at `publish_diagnostics` is too late. Perhaps we should ignore these files in `generate_lint_diagnostics` so that we don't bother doing this work when not needed?

Ready.

> Yes, I believe that would be better. Also, it looks like we're not handling `files.include` in addition to `files.exclude`.

Handling of `files.include` should be after excluding?

Reply:
Seems handling of `include` section has no meaning.
For `deno lint` it points to external files and does not re-include excluded files, but LSP (I've tested with VSCode) always lints all the opened files in an editor, even if they are from another disk partition (bug?), so, this section is useless. Correct me if I'm wrong.

Reply:
With `files.include`, if it is not empty I believe it would ignore any files/folders not in `documents` if it's not in `files.include`.

Reply:
>it would ignore any files/folders not in `documents`

Sorry, I don't sure I understand, but `generate_lint_diagnostics` collects diagnostics only from `documents`.
Maybe You meant:
> if `files.include` is not empty - ignore anything that is not in `files.include`

Reply:
Sorry! Yes. I fixed my sentence.